feat: Introduce `enforce-hash-update` flag to help hashing migration (#10209)
serverless/serverless: docs/guides/upgrading-v3.md

`alexaSkill` events now require an `appId` ([learn more](../deprecations.md#support-for-alexaskill-event-without-appid-is-to-be-removed)). That change was required to implement a more stable deployment, as well as to deploy more restricted IAM permissions.

### Lambda Hashing Algorithm

By default, Lambda version hashes will now be generated using improved algorithm (fixed determinism issues). As it is a breaking change that requires more manual effort during migration, it is still possible (but not recommended) to keep using old algorithm by using the following configuration:

> ````

Reviewer: I think it should be one back quote less

`alexaSkill` events now require an `appId` ([learn more](../deprecations.md#support-for-alexaskill-event-without-appid-is-to-be-removed)). That change was required to implement a more stable deployment, as well as to deploy more restricted IAM permissions.

### Lambda Hashing Algorithm

> By default, Lambda version hashes will now be generated using improved algorithm (fixed determinism issues). As it is a breaking change that requires more manual effort during migration, it is still possible (but not recommended) to keep using old algorithm by using the following configuration:

Reviewer 2: ```suggestion
By default, Lambda version hashes will now be generated using an improved algorithm (fixes determinism issues). As it is a breaking change that requires more manual effort during migration, it is still possible (but not recommended) to keep using the old algorithm by using the following configuration:
```

```
provider:
  lambdaHashingVersion: 20200924
```

> However, we highly encourage an upgrade to the new algorithm. Safe upgrade requires change in configuration or code of all your functions. You can do it safely following the guide below:

Reviewer 2: "Safe upgrade" makes me think (as a user) that something may not be safe (loosing stuff? destroying resources accidentally?)

I'd rephrase with something like:

```suggestion
However, we highly encourage an upgrade to the new algorithm. To upgrade, you must redeploy your service with code or configuration changes in all functions. You can do it by following the guide below:
```

**NOTE**: Please keep in mind that these changes require two deployments with manual configuration adjustment between them. It also creates two additional versions and temporarily overrides descriptions of your functions. Migration will need to be done separately for each of your environments/stages.

> 1. Run `sls deploy` with additional `--enforce-hash-update` flag to override description for all your functions and force creation of new versions.

Reviewer 2: ```suggestion
1. Run `sls deploy` with additional `--enforce-hash-update` flag: that flag will override the description for all Lambda functions, which will force the creation of new versions.
```

I would also state clearly that we start from a Serverless v2 application. Maybe it's a bit over the top to explicitly write that? But at least as a user it's 100% clear. Something like this:

```
1. In a Serverless Framework v2 project
```

(or maybe put that part in the paragraph above?)

Author: I think we can add it, but I think it's also a bit redundant since this guide is about migrating from v2 to v3 specifically. Where would you exactly see a good place to add such information?

Reviewer 2: 👍  no problem then ignore that part

> 2. Set `provider.lambdaHashingVersion` to `20201221` in your configuration.

Reviewer 2: ```suggestion
2. Set `provider.lambdaHashingVersion` to `20201221` in your configuration: your service will now always deploy with the new Lambda version hashes (which is the new default in v3).
```

(just so that users understand what this step does in regards to v2 and v3)

> 3. Run `sls deploy`, this time without additional `--enforce-hash-update` flag.

Reviewer 2: ```suggestion
3. Run `sls deploy`, this time without additional `--enforce-hash-update` flag: that will restore the original descriptions on all Lambda functions.
```

That will reassure users that things go back to normal.

> Now your whole service will be fully migrated to new Lambda Hashing Algorithm.

Reviewer 2: ```suggestion
Now your whole service is fully migrated to the new Lambda Hashing Algorithm.
```
(the present tense was used in the sentences above)

      statements: ...
      managedPolicies: ...
    deploymentRole: ...

> ````

Reviewer: And then this change should not happen
